# Notebook 17 – Mini Assessment

In [1]:
import pandas as pd
df = pd.read_csv("hr_employee_attrition_raw.csv")
df.head()

,EmployeeID,Age,Gender,Department,JobRole,Education,MonthlyIncome,YearsAtCompany,JobSatisfaction,PerformanceRating,DistanceFromHomeKM,OverTime,JoinDate,Attrition,EmployeeCountFlag,RandomSurveyCode
0,EMP10507,23,Female,Finance,Accountant,Bachelor's,2227.56,8,3.0,1,1.8,No,2019-07-23,No,1,448909
1,EMP10740,29,Male,Finance,Accountant,High School,5558.6,15,1.0,3,5.7,Yes,2022-02-16,No,1,724794
2,EMP10278,45,Female,sales,Brand Manager,NaN,6167.63,3,4.0,1,16.7,No,2016-07-16,Yes,1,571964
3,EMP10929,35,Male,HR,HR Manager,Bachelor's,5492.97,0,1.0,2,0.8,Yes,2024-08-19,No,1,397616
4,EMP10632,34,Male,Support,Support Lead,Bachelor's,5433.06,1,NaN,3,9.6,No,2015-03-23,No,1,441126




**Q1. What is data preprocessing?**
Cleaning and transforming raw data into a format a machine learning model can use effectively.

**Q2. Why is preprocessing required?**
Raw data usually has errors, missing values, wrong types, and inconsistent scales, models can't learn correctly from that.

**Q3. Explain MCAR, MAR, MNAR.**
MCAR: missingness is totally random, unrelated to any data. MAR: missingness relates to another observed column (e.g., income missing more often for a specific department). MNAR: missingness relates to the value itself (e.g., people with very high salaries refusing to disclose it).

**Q4. How would you handle 40% missing values?**
Depends on the column's importance and missingness pattern. If it's a critical feature, impute carefully or use a "missing" flag category. If it's low-value or the missingness is systematic (MNAR), consider dropping the column entirely.

**Q5. Mean vs Median Imputation.**
Mean works well for roughly symmetric, outlier-free data. Median is safer when data is skewed or has outliers, since it isn't pulled by extreme values.

**Q6. What is data leakage?**
When information that wouldn't be available at prediction time accidentally influences training, causing misleadingly high performance.

**Q7. Explain target leakage.**
Using a feature that only exists or has meaning because of the outcome itself, e.g., an "exit interview completed" flag that only exists for employees who already left.

**Q8. What is one-hot encoding?**
Converting a categorical column into multiple binary (0/1) columns, one per category, so no false numeric order is implied.

**Q9. Label Encoding vs One-Hot Encoding.**
Label encoding assigns arbitrary integers, appropriate for ordinal data or tree models. One-hot encoding creates separate binary columns, appropriate for nominal data to avoid implying false order.

**Q10. Standardization vs Normalization.**
Standardization rescales to mean 0, std 1 (unbounded range). Normalization rescales into a fixed range, usually 0 to 1.

**Q11. When would you use RobustScaler?**
When the column has outliers you haven't removed yet, since it uses median and IQR instead of mean and min/max.

**Q12. What is class imbalance?**
When one class in the target variable vastly outnumbers the other(s), e.g., 81% "No" vs 19% "Yes" attrition.

**Q13. Explain SMOTE.**
Generates new synthetic minority-class samples by interpolating between existing minority points and their nearest neighbors, instead of just duplicating rows.

**Q14. Why fit preprocessing only on training data?**
So no information from the test set (mean, categories, frequencies) leaks into training, which would produce an unrealistically optimistic evaluation.

**Q15. What is a preprocessing pipeline?**
A structured, ordered sequence of preprocessing steps (imputation, scaling, encoding) bundled into a single reusable object.

**Q16. Why use ColumnTransformer?**
It lets you apply different preprocessing steps to different columns (e.g., scaling for numeric, encoding for categorical) in one unified step, rather than juggling separate manual transformations.